Feature/gitlab ci

rmelotte opened the pull request:
Closes #504.
We would like to switch from Travis to Gitlab CI because it's a lot more flexible.

This PR adds a first .gitlab-ci.yml file, which more or less reproduces the Travis configuration. 
Once merged, it will then be synced to the existing [mirror](https://git.prpl.dev/prplmesh/prplMesh).
The content of this PR has already been tested on Gitlab CI, the results are [here](https://git.prpl.dev/prplmesh/prplMesh/pipelines/61).

rmelotte committed:
Add a .gitlab-ci.yml file based on the travis.yml

A .gitlab-ci.yml that more or less reproduces the travis.yml
configuration was added.

Running the tests requires docker in the tests images. Some of the
possible ways to do that with Gitlab CI is to either:
- use a shell executor.
- use the docker-in-docker workflow with the docker executor.
- bind-mount the docker socket inside the test image.

For reference, this is explained here:
https://docs.gitlab.com/ee/ci/docker/using_docker_build.html

Because it was the most straightforward for us to implement, we chose
the first solution. However, as the reproducibility of the shell
runner is not so good, it might be worth it to switch to the
docker-in-docker workflow at some point.

Signed-off-by: Raphaël Mélotte <raphael.melotte@essensium.com>
Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

rmelotte committed:
tools: docker: remove the "tty" and "interactive" options

The "--tty" and "--interactive" options have been removed from most of
the test scripts because:
- they are not needed to run the tests.
- they prevent the scripts to be used with Gitlab CI.

Signed-off-by: Raphaël Mélotte <raphael.melotte@essensium.com>

tomereli commented on .gitlab-ci.yml:
```diff
@@ -0,0 +1,29 @@
+  stage: test
+  script:
+    - cd tools/docker
+    - ./run.sh -f -d --entrypoint "$PWD/../../../build/install/bin/tests/tlvf_test"
```
What about encryption test?

arnout replied:
It is currently not in .travis.yml. I think it's a good idea for the first version of .gitlab-ci.yml to be as close as possible to .travis.yml.

arnout commented:
> I think thus will mean that running test_gateway_repeater.sh will now kill the container after it is complete? If not, will we still be able to run docker exec - it gateway bash and debug as we normally do?

`--interactive --tty` is completely unrelated to killing the container.

`--interactive` will make sure that stdin is connected. So it is completely useless on any run or exec command that doesn't actually use stdin.

`--tty` will allocate a pty. In practice, this means that isatty() will return true, which will make some programs behave differently. For example, there won't be any colours in the output of `make`.

That the containers keep on running is proven by the fact that Travis CI succeeded on this PR: test_flows.sh requires the gateway, repeater1 and repeater2 containers to be running...

arnout commented on .gitlab-ci.yml:
```diff
@@ -0,0 +1,29 @@
+before_script:
+  - docker info
```
Nice for your debugging, but is it useful in production?

rmelotte replied:
It's useful to have when a test fails, but as in our case we will always have access to the runners it's probably not needed. I'll remove it.

arnout commented on .gitlab-ci.yml:
```diff
@@ -0,0 +1,29 @@
+  script:
+    - cd tools/docker
+    # Try to pull from dockerhub, build the container images if that fails
+    - ./image-pull.sh || ./image-build.sh
```
For building, it should be possible to use the prplmesh-builder:ubuntu-18.04 container directly. Of course we can't use the shell runner then, so remove the shell tag.

Note that this is something we should probably do in a follow-up PR - for this PR, it's good to stay as close as possible to .travis.yml.

D'oh, I realize now that the builder container has maptools as the entrypoint, so that won't work... We'd first have to remove the entrypoint from that container so we have full flexibility for running stuff in it.

arnout commented on .gitlab-ci.yml:
```diff
@@ -0,0 +1,29 @@
+    # Try to pull from dockerhub, build the container images if that fails
+    - ./image-pull.sh || ./image-build.sh
+    # Run clang-format and check there are no modified files by checking git status
+    - ./clang-format.sh
```
We can use the top-level clang-format.sh script itself if we run it in the appropriate docker container.

It would also be nice to put this in a separate job so it can be done in parallel (of course the git diff-index has to be included in the same job).

Again, both of these comments are for follow-up PRs.

arnout commented on .gitlab-ci.yml:
```diff
@@ -0,0 +1,29 @@
+  script:
+    - cd tools/docker
+    - ./run.sh -f -d --entrypoint "$PWD/../../../build/install/bin/tests/tlvf_test"
+    - ./tests/test_flows.sh -v
```
This will leave the test containers running directly in the machine which has the runner (i.e., mind2). Not nice... Please add a after_script to kill the containers. This **is** something that must be fixed for this PR.

Also, since this is in a shell runner, the containers started by different instances all have the same name "gateway", "repeater1", "repeater2". Therefore, if two jobs are executing concurrently, they'll conflict with each other. We need to find a way to resolve that - but again, can be done later.

arnout commented on .gitlab-ci.yml:
```diff
@@ -0,0 +1,29 @@
+    - git update-index -q --refresh
+    - git diff-index --exit-code HEAD
+  tags:
+    - shell
```
Please use the "dockerbuild" tag. If we later switch to one of the other approaches, the tag will stay the same, it will just be a differently-configured runner.

arnout reviewed (approved):
Lots of comments but mostly for follow-up PRs. Just two things that need to be fixed: "dockerbuild" tag (for both) and an after_script to kill the containers.

rmelotte committed:
tools: docker: add a way to stop the containers

Currently, the test scripts each start a number of containers, but
there is no easy way to stop them all.

For now, their name are known and hard-coded so they can be stopped
manually by their name, but that may change in the near future (to be
able to run multiple tests in parallel for example, we need unique
container names).

For this reason, this commit does two things:
- make tools/docker/run.sh add the name of every container it runs to
  a '.test_containers' file.
- add the stop.sh script, which allows to stop, kill, or remove
  containers that were started through run.sh (which all current tests
  scripts do).

Also, a .gitignore file with '.test_containers' is added to the docker
folder.

Signed-off-by: Raphaël Mélotte <raphael.melotte@essensium.com>

rmelotte commented:
I made the required changes, and opened an issue (#515) for the comments that won't be solved as part of this PR.
The new Gitlab pipeline is [here](https://git.prpl.dev/prplmesh/prplMesh/pipelines/63).

rmelotte commented:
I forgot to remove the "docker info" part, which is now done. 
Pipeline: https://git.prpl.dev/prplmesh/prplMesh/pipelines/64
